src/NHibernate/Linq/NhLinqDmlExpression.cs

```diff
@@ -17,6 +17,10 @@ public NhLinqDmlExpression(QueryMode queryMode, Expression expression, ISessionF
 		{
 			Key = $"{queryMode.ToString().ToUpperInvariant()} {Key}";
 			QueryMode = queryMode;
+
+			//For DML queries parameters check doesn't really work (queries are either not cached or cached incorrectly - see GH-2222 and GH-2298)
+			//So for now disable caching for DML queries
+			CanCachePlan = false;
```

Comment:
The real fix for #2298 is in `ExpandedQueryExpression`. But as long DML CanCachePlan check is not functional it's better to disable it completely.

Reply:
I still have to do more checks, but I now think we have a combination of bugs:
 - `NhLinqExpression` may be wrapped inside `ExpandedQueryExpression`, causing their "custom" cache disabling logic to be defeated in `QueryPlanCache`. This PR fix this point, in a good way I think
 - Projections seem not included in the query plan cache key at least in the DML case, causing different projections to be considered equal. This change I am commenting is a work-around for this (including the private to protected change). I think it should be handled in a dedicated PR.
 - "Thanks" to #2222, the previous trouble is hidden when expression are not expanded.

Since the trouble only occurs with queries using parameter lists, and is not a regression, I am not convinced we should issue patches for it. A fix in 5.3 would be enough IMO.
If fixing in 5.3, better fix it with two PR, one for properly handling the cache disabling logic as this PR does, another one for handling the projection cache key trouble.
Otherwise, if still fixing in 5.2 and backporting all the way down to 5.0, ok for keeping the workaround of disabling the cache here.